loki/transformations/data_offload.py

```diff
@@ -908,3 +924,160 @@ def _append_routine_arguments(self, routine, item):
             )) for arg in new_arguments
         ]
         routine.arguments += tuple(sorted(new_arguments, key=lambda symbol: symbol.name))
+
+
+def find_target_calls(region, targets):
+    """Returns a list of all calls to targets inside the region
+
+    Parameters
+    ----------
+    :region: :any:`PragmaRegion`
+    :targets: collection of :any:`Subroutine`
+        Iterable object of subroutines or functions called
+    :returns: list of :any:`CallStatement`
+    """
+    calls = FindNodes(CallStatement).visit(region)
+    calls = [c for c in calls if str(c.name).lower() in targets]
+    return calls
+
+
+class FieldOffloadTransformation(Transformation):
+    class FieldPointerMap:
+        def __init__(self, devptrs, inargs, inoutargs, outargs):
+            self.inargs = inargs
+            self.inoutargs = inoutargs
+            self.outargs = outargs
+            self.devptrs = devptrs
+
+        @property
+        def in_pairs(self):
+            for i, inarg in enumerate(self.inargs):
+                yield inarg, self.devptrs[i]
+
+        @property
+        def inout_pairs(self):
+            start = len(self.inargs)
+            for i, inoutarg in enumerate(self.inoutargs):
+                yield inoutarg, self.devptrs[i+start]
+
+        @property
+        def out_pairs(self):
+            start = len(self.inargs)+len(self.inoutargs)
+            for i, outarg in enumerate(self.outargs):
+                yield outarg, self.devptrs[i+start]
+
+
+    def __init__(self, **kwargs):
+        self.deviceptr_prefix = kwargs.get('devptr_prefix', 'loki_devptr_')
+        field_group_types = kwargs.get('field_group_types', ['CLOUDSC_STATE_TYPE',
+                                                             'CLOUDSC_AUX_TYPE',
+                                                             'CLOUDSC_FLUX_TYPE'])
+        self.field_group_types = tuple(typename.lower() for typename in field_group_types)
+        self.offload_index = kwargs.get('offload_index', 'IBL')
+
+    def transform_subroutine(self, routine, **kwargs):
+        role = kwargs['role']
+        targets = as_tuple(kwargs.get('targets'), (None))
+        if role == 'driver':
+            self.process_driver(routine, targets)
+
+    def process_driver(self, driver, targets):
+        remove_field_api_view_updates(driver, self.field_group_types + tuple(s.upper() for s in self.field_group_types))
```

Comment:
The lower/upper-case conversion here should not be necessary. If `remove_field_api_view_updates` is not sufficiently case-agnostic, it should be fixed there.